docs/P4c-test-documentation.md

```diff
@@ -0,0 +1,235 @@
+# **Adding New Tests to the P4C Repository**
+
+**This document provides a guide for developers who want to add new tests to the P4C repository. It explains where to place test files, how to manage expected output files, and how to handle tests that are expected to fail (Xfail).**
+
+---
+
+
+
+
+
+## **1. Adding "Positive" Tests**
+
+Positive tests are those where the P4 compiler should process the input without any errors.
+
+### Steps for Adding Positive Tests
+
+1. **Source File Location**:
+
+   - Positive tests are located in multiple directories based on the purpose and the backend/tool being tested. Below is a description of the primary directories under `p4c/testdata` and their purposes:
+
+     - **`p4_16_samples`**: Contains general positive tests for the P4\_16 language.
+     - **`p4_16_samples_outputs`**: Stores the expected outputs for the tests in `p4_16_samples`.
+     - **`p4_14_samples`**: Contains tests specific to the older P4\_14 language version.
+     - **`p4_14_samples_outputs`**: Stores the expected outputs for tests in `p4_14_samples`.
+     - **`bmv2`**: Holds tests targeting the BMv2 backend, including packet-processing logic and examples.
+     - **`psa`**: Includes tests for the PSA (Portable Switch Architecture) target.
+     - **`ebpf`**: Contains tests designed for the eBPF backend.
+     - **`xdp`**: Stores tests related to XDP programs using eBPF.
+     - **`arch`**: Contains architecture-specific tests, covering various P4 architectures.
+     - **`tools`**: Includes tests for tools provided in the P4C repository, such as `p4test` and `p4c-bm2-ss`.
+     - **`error_samples`**: Contains tests that verify the compiler's behavior when encountering erroneous P4 programs.
+     - **`stf_samples`**: Stores `.stf` (Switch Test Framework) files associated with P4 programs, defining packet inputs/outputs for testing.
+     - **`stress_tests`**: Contains larger or more computationally intensive tests to stress-test the compiler.
+
+   - Make sure the file you are adding is placed in the appropriate directory based on the purpose and backend/tool it is testing.
+   - Ensure the file is named descriptively to reflect its purpose (e.g., `example_positive_test.p4`).
+
+
+2. **File Naming Conventions**:
+
+   - For **BMv2-specific tests** with packets, the source file name must match the pattern `*-bmv2.p4`.
+  - For **v1model architecture** tests: `*-bmv2-v1model.p4`.
+
+  - For **PSA architecture** tests: `*-bmv2-psa.p4`.
+  
+   - Additional file name patterns:
+     - For eBPF-specific tests: `*_ebpf.p4` or `*_ebpf-kernel.p4`.
+     - For graph backend tests: `*graph*.p4`.
+     - For UBPF-specific tests: `*_ubpf.p4`.
+
+     - Other specialized tests under subdirectories such as:
+       - `p4tc_samples/` for TC tests.
+       - `fabric_*/` and `omec/` for Fabric and OMEC tests.
```

Comment:
I am starting to get redundant with my comments, but in general, I think it is important the developers know the reason for why these directories are significant, and what causes the set of tests to be created, that are created.  Again, it is because of the contents of one or more CMakeLists.txt files.  For the `fabric_*' directories it is because of these lines, I think:
```
./backends/p4test/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/fabric_*/fabric.p4"
./backends/bmv2/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/fabric_*/fabric.p4"
```